File: document/doc/reference/operation/d8_flow_direction/example-1/d8_flow_direction.py
import sys
from pathlib import Path

import docopt

import lue.document as ld
import lue.framework as lfr
from lue import __version__ as version


@lfr.runtime_scope
def main() -> None:
    command = Path(sys.argv[0]).name
    usage = f"""\
Usage:
    {command} <elevation> <flow_direction>

Options:
    -h --help         Show this screen and exit
    --version         Show version and exit
    <elevation>       Elevation array to read
    <flow_direction>  Flow direction array to write
"""
    arguments = docopt.docopt(usage, sys.argv[1:], version=version)
    elevation_array_path = Path(arguments["<elevation>"])
    flow_direction_array_path = Path(arguments["<flow_direction>"])

    elevation = ld.read_array(elevation_array_path)

    # [example
    flow_direction = lfr.d8_flow_direction(elevation)
    # example]

    lfr.to_gdal(flow_direction, str(flow_direction_array_path))


if __name__ == "__main__":
    sys.exit(main())

File: document/doc/reference/operation/routing/d8_flow_direction/example-1/d8_flow_direction.cpp
#include "lue/framework/algorithm/value_policies/d8_flow_direction.hpp"
#include "lue/framework/algorithm/value_policies/add.hpp"
#include "lue/document.hpp"
#include "lue/framework.hpp"
#include <hpx/hpx_main.hpp>


class Example: public lue::document::Example
{

    public:

        Example(int const argc, char const* const* argv):

            lue::document::Example{
                [&]() -> cxxopts::Options
                {
                    cxxopts::Options options{std::filesystem::path{argv[0]}.filename()};

                    options.positional_help("<elevation> <flow_direction>");
                    options.add_options()
                        // clang-format off
                        ("elevation_array", "array to read", cxxopts::value<std::string>())
                        ("flow_direction_array", "array to write", cxxopts::value<std::string>())
                        // clang-format on
                        ;
                    options.parse_positional({"elevation_array", "flow_direction_array"});

                    return options;
                }(),
                argc,
                argv}

        {
        }

    private:

        auto run_implementation() -> int override
        {
            if (!argument_parsed("elevation_array") || !argument_parsed("flow_direction_array"))
            {
                throw std::runtime_error("Wrong usage, pass '--help' for more information");
            }

            auto const elevation_array_pathname = argument<std::string>("elevation_array");
            auto const flow_direction_array_pathname = argument<std::string>("flow_direction_array");

            using namespace lue;
            using namespace value_policies;
            Rank const rank{2};

            using FloatElement = FloatingPointElement<0>;
            using FloatArray = PartitionedArray<FloatElement, rank>;
            using FlowDirectionArray = PartitionedArray<FlowDirectionElement, rank>;

            FloatArray const elevation = read_array<FloatElement, rank>(elevation_array_pathname);

            // [example
            FlowDirectionArray flow_direction = d8_flow_direction<FlowDirectionElement>(elevation);
            // example]

            to_gdal(flow_direction, flow_direction_array_pathname);

            return EXIT_SUCCESS;
        }
};


auto main(int argc, char** argv) -> int
{
    return Example{argc, argv}.run();
}
